packages/playwright-core/src/server/frames.ts

```diff
@@ -488,6 +488,7 @@ export class Frame extends SdkObject {
   }
 
   setPendingDocument(documentInfo: DocumentInfo | undefined) {
+    this._page._frameManager._dialogCounter = 0;
```

Comment:
I think we should listen to `Page.javascriptDialogClosed` instead - that will also help when the user manually closes the dialog in headed mode.